File: applications/admin-panel/src/services/bridgeApi.ts
/**
 * Bridge API Client
 *
 * Communicates with the ArmorClaw bridge via HTTP proxy to Unix socket.
 * All requests go through /api/* which is proxied to the bridge.
 */

// Types
export interface LockdownStatus {
  mode: 'lockdown' | 'bonding' | 'configuring' | 'hardening' | 'operational';
  admin_established: boolean;
  single_device_mode: boolean;
  allowed_communication: string[];
  setup_complete: boolean;
  security_configured: boolean;
  keystore_initialized: boolean;
  secrets_injected: boolean;
  hardening_complete: boolean;
}

export interface DataCategory {
  id: string;
  name: string;
  description: string;
  examples: string[];
  risk_level: 'high' | 'medium' | 'low';
  permission: 'deny' | 'allow' | 'allow_all';
  allowed_websites: string[];
  requires_approval: boolean;
}

export interface SecurityTier {
  id: string;
  name: string;
  description: string;
  restrictions: string[];
}

export interface Device {
  id: string;
  name: string;
  type: 'phone' | 'tablet' | 'desktop';
  platform: string;
  trust_state: 'verified' | 'unverified' | 'pending_approval' | 'rejected';
  last_seen: string;
  first_seen: string;
  ip_address: string;
  user_agent: string;
  is_current: boolean;
}

export interface Adapter {
  id: string;
  name: string;
  description: string;
  status: 'connected' | 'disconnected' | 'error' | 'pending';
  enabled: boolean;
  permissions: string[];
  last_sync?: string;
  error?: string;
}

export interface Invite {
  id: string;
  code: string;
  role: 'admin' | 'moderator' | 'user';
  created_by: string;
  created_at: string;
  expires_at?: string;
  max_uses: number;
  use_count: number;
  status: 'active' | 'used' | 'expired' | 'revoked' | 'exhausted';
}

export interface APIKey {
  id: string;
  provider: string;
  name: string;
  prefix: string;
  created_at: string;
  last_used?: string;
  status: 'active' | 'revoked' | 'expired';
}

export interface AuditEvent {
  id: string;
  timestamp: string;
  category: 'security' | 'auth' | 'data' | 'adapter' | 'system';
  action: string;
  actor: string;
  resource: string;
  outcome: 'success' | 'failure' | 'warning';
  details: string;
  ip_address: string;
  device_id?: string;
}

export interface ClaimRequest {
  matrix_user_id: string;
  device_id: string;
  user_agent: string;
  ip_address: string;
  device_name: string;
}

export interface ClaimResponse {
  id: string;
  token: string;
  challenge: string;
  challenge_type: string;
  status: string;
  created_at: string;
  expires_at: string;
}

export interface QRResult {
  token: {
    id: string;
    token: string;
    type: string;
    state: string;
    expires_at: string;
  };
  url: string;
  deep_link: string;
  qr_image: string; // Base64 encoded
  expires_at: string;
}

// RPC Request/Response types
interface RPCRequest {
  jsonrpc: '2.0';
  id: number;
  method: string;
  params?: Record<string, unknown>;
}

interface RPCResponse<T> {
  jsonrpc: '2.0';
  id: number;
  result?: T;
  error?: {
    code: number;
    message: string;
    data?: unknown;
  };
}

// API Client
class BridgeAPIClient {
  private requestId = 0;
  private baseUrl = '/api';

  async rpc<T>(method: string, params?: Record<string, unknown>): Promise<T> {
    this.requestId++;

    const request: RPCRequest = {
      jsonrpc: '2.0',
      id: this.requestId,
      method,
      params,
    };

    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      throw new Error(`HTTP error: ${response.status}`);
    }

    const data: RPCResponse<T> = await response.json();

    if (data.error) {
      throw new Error(data.error.message);
    }

    return data.result as T;
  }

  // Lockdown methods
  async getLockdownStatus(): Promise<LockdownStatus> {
    return this.rpc<LockdownStatus>('lockdown.status');
  }

  async getChallenge(): Promise<{ nonce: string; expires_at: string }> {
    return this.rpc('lockdown.get_challenge');
  }

  async claimOwnership(params: {
    display_name: string;
    device_name: string;
    device_fingerprint: string;
    passphrase_commitment: string;
    challenge_response?: string;
  }): Promise<{
    status: string;
    admin_id: string;
    device_id: string;
    certificate: string;
    session_token: string;
    expires_at: string;
    next_step: string;
  }> {
    return this.rpc('lockdown.claim_ownership', params);
  }

  async transitionMode(target: string): Promise<{ success: boolean; mode: string }> {
    return this.rpc('lockdown.transition', { target });
  }

  // Security methods
  async getSecurityCategories(): Promise<DataCategory[]> {
    return this.rpc<DataCategory[]>('security.get_categories');
  }

  async setSecurityCategory(category: string, permission: string): Promise<{ success: boolean }> {
    return this.rpc('security.set_category', { category, permission });
  }

  async getSecurityTiers(): Promise<SecurityTier[]> {
    return this.rpc<SecurityTier[]>('security.get_tiers');
  }

  async setSecurityTier(tier: string): Promise<{ success: boolean }> {
    return this.rpc('security.set_tier', { tier });
  }

  // Device methods
  async listDevices(): Promise<Device[]> {
    return this.rpc<Device[]>('device.list');
  }

  async getDevice(deviceId: string): Promise<Device> {
    return this.rpc('device.get', { device_id: deviceId });
  }

  async approveDevice(deviceId: string, approvedBy: string): Promise<{ success: boolean }> {
    return this.rpc('device.approve', { device_id: deviceId, approved_by: approvedBy });
  }

  async rejectDevice(deviceId: string, rejectedBy: string, reason: string): Promise<{ success: boolean }> {
    return this.rpc('device.reject', { device_id: deviceId, rejected_by: rejectedBy, reason });
  }

  // Adapter methods
  async listAdapters(): Promise<Adapter[]> {
    return this.rpc<Adapter[]>('adapter.list');
  }

  async enableAdapter(adapterId: string): Promise<{ success: boolean }> {
    return this.rpc('adapter.enable', { adapter_id: adapterId });
  }

  async disableAdapter(adapterId: string): Promise<{ success: boolean }> {
    return this.rpc('adapter.disable', { adapter_id: adapterId });
  }

  async configureAdapter(adapterId: string, config: Record<string, unknown>): Promise<{ success: boolean }> {
    return this.rpc('adapter.configure', { adapter_id: adapterId, config });
  }

  // Invite methods
  async createInvite(params: {
    role: string;
    expiration: string;
    max_uses: number;
    welcome_message?: string;
    created_by: string;
  }): Promise<Invite> {
    return this.rpc('invite.create', params);
  }

  async listInvites(): Promise<Invite[]> {
    return this.rpc<Invite[]>('invite.list');
  }

  async revokeInvite(inviteId: string, revokedBy: string): Promise<{ success: boolean }> {
    return this.rpc('invite.revoke', { invite_id: inviteId, revoked_by: revokedBy });
  }

  async validateInvite(code: string): Promise<Invite> {
    return this.rpc('invite.validate', { code });
  }

  // QR methods
  async generateSetupQR(): Promise<QRResult> {
    return this.rpc('qr.generate_setup');
  }

  async generateBondingQR(challenge: string): Promise<QRResult> {
    return this.rpc('qr.generate_bonding', { challenge });
  }

  async generateSecretQR(provider: string): Promise<QRResult> {
    return this.rpc('qr.generate_secret', { provider });
  }

  async generateInviteQR(inviteCode: string, role: string): Promise<QRResult> {
    return this.rpc('qr.generate_invite', { invite_code: inviteCode, role });
  }

  // Admin claim methods (for Element X integration)
  async initiateAdminClaim(params: ClaimRequest): Promise<ClaimResponse> {
    return this.rpc('admin.initiate_claim', params);
  }

  async validateAdminToken(token: string): Promise<ClaimResponse> {
    return this.rpc('admin.validate_token', { token });
  }

  async respondChallenge(token: string, response: string): Promise<{ success: boolean }> {
    return this.rpc('admin.respond_challenge', { token, response });
  }

  async approveClaim(claimId: string, approvedBy: string): Promise<{ success: boolean }> {
    return this.rpc('admin.approve_claim', { claim_id: claimId, approved_by: approvedBy });
  }

  async rejectClaim(claimId: string, rejectedBy: string, reason: string): Promise<{ success: boolean }> {
    return this.rpc('admin.reject_claim', { claim_id: claimId, rejected_by: rejectedBy, reason });
  }

  // Audit methods
  async getAuditLog(params?: {
    category?: string;
    search?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ events: AuditEvent[]; total: number }> {
    return this.rpc('audit.get_log', params);
  }
}

// Export singleton instance
export const bridgeApi = new BridgeAPIClient();

// React Query hooks
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';

// Lockdown hooks
export function useLockdownStatus() {
  return useQuery({
    queryKey: ['lockdown', 'status'],
    queryFn: () => bridgeApi.getLockdownStatus(),
    refetchInterval: 5000, // Refresh every 5 seconds during setup
  });
}

// Security hooks
export function useSecurityCategories() {
  return useQuery({
    queryKey: ['security', 'categories'],
    queryFn: () => bridgeApi.getSecurityCategories(),
  });
}

export function useSetSecurityCategory() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ category, permission }: { category: string; permission: string }) =>
      bridgeApi.setSecurityCategory(category, permission),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['security', 'categories'] });
    },
  });
}

// Device hooks
export function useDevices() {
  return useQuery({
    queryKey: ['devices'],
    queryFn: () => bridgeApi.listDevices(),
  });
}

export function useApproveDevice() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ deviceId, approvedBy }: { deviceId: string; approvedBy: string }) =>
      bridgeApi.approveDevice(deviceId, approvedBy),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['devices'] });
    },
  });
}

// Adapter hooks
export function useAdapters() {
  return useQuery({
    queryKey: ['adapters'],
    queryFn: () => bridgeApi.listAdapters(),
  });
}

export function useToggleAdapter() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ adapterId, enabled }: { adapterId: string; enabled: boolean }) =>
      enabled
        ? bridgeApi.enableAdapter(adapterId)
        : bridgeApi.disableAdapter(adapterId),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['adapters'] });
    },
  });
}

// Invite hooks
export function useInvites() {
  return useQuery({
    queryKey: ['invites'],
    queryFn: () => bridgeApi.listInvites(),
  });
}

export function useCreateInvite() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (params: Parameters<typeof bridgeApi.createInvite>[0]) =>
      bridgeApi.createInvite(params),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['invites'] });
    },
  });
}

// QR hooks
export function useSetupQR() {
  return useQuery({
    queryKey: ['qr', 'setup'],
    queryFn: () => bridgeApi.generateSetupQR(),
    staleTime: 60000, // QR valid for 1 minute
  });
}

// Audit hooks
export function useAuditLog(params?: { category?: string; search?: string }) {
  return useQuery({
    queryKey: ['audit', params],
    queryFn: () => bridgeApi.getAuditLog(params),
  });
}

File: applications/ArmorChat/app/src/main/java/app/armorclaw/network/BridgeApi.kt
package app.armorclaw.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Bridge API Client for ArmorChat
 *
 * Communicates with the ArmorClaw bridge via HTTP.
 * All requests are JSON-RPC 2.0.
 */
class BridgeApi(private val baseUrl: String = "http://localhost:8080/api") {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private var requestId = 0

    //region Types

    @Serializable
    data class RPCRequest(
        val jsonrpc: String = "2.0",
        val id: Int,
        val method: String,
        val params: Map<String, String>? = null
    )

    @Serializable
    data class RPCResponse<T>(
        val jsonrpc: String,
        val id: Int,
        val result: T? = null,
        val error: RPCError? = null
    )

    @Serializable
    data class RPCError(
        val code: Int,
        val message: String,
        val data: String? = null
    )

    @Serializable
    data class LockdownStatus(
        val mode: String,
        val admin_established: Boolean,
        val single_device_mode: Boolean,
        val setup_complete: Boolean,
        val security_configured: Boolean,
        val keystore_initialized: Boolean
    )

    @Serializable
    data class Challenge(
        val nonce: String,
        val created_at: String,
        val expires_at: String
    )

    @Serializable
    data class BondingRequest(
        val display_name: String,
        val device_name: String,
        val device_fingerprint: String,
        val passphrase_commitment: String,
        val challenge_response: String? = null
    )

    @Serializable
    data class BondingResponse(
        val status: String,
        val admin_id: String,
        val device_id: String,
        val certificate: String,
        val session_token: String,
        val expires_at: String,
        val next_step: String
    )

    @Serializable
    data class DataCategory(
        val id: String,
        val name: String,
        val description: String,
        val risk_level: String,
        val permission: String,
        val allowed_websites: List<String>,
        val requires_approval: Boolean
    )

    @Serializable
    data class SecurityTier(
        val id: String,
        val name: String,
        val description: String
    )

    @Serializable
    data class Device(
        val id: String,
        val name: String,
        val type: String,
        val trust_state: String,
        val last_seen: String,
        val is_current: Boolean
    )

    @Serializable
    data class Invite(
        val id: String,
        val code: String,
        val role: String,
        val status: String,
        val created_at: String,
        val expires_at: String?
    )

    //endregion

    //region API Methods

    /**
     * Generic RPC call
     */
    private inline fun <reified T> rpc(method: String, params: Map<String, String>? = null): Result<T> {
        return try {
            requestId++
            val request = RPCRequest(id = requestId, method = method, params = params)
            val requestBody = json.encodeToString(RPCRequest.serializer(), request)
                .toRequestBody("application/json".toMediaType())

            val httpRequest = Request.Builder()
                .url(baseUrl)
                .post(requestBody)
                .build()

            val response = client.newCall(httpRequest).execute()
            if (!response.isSuccessful) {
                return Result.failure(Exception("HTTP ${response.code}: ${response.message}"))
            }

            val responseBody = response.body?.string()
                ?: return Result.failure(Exception("Empty response body"))

            val rpcResponse = json.decodeFromString<RPCResponse<T>>(responseBody)

            if (rpcResponse.error != null) {
                Result.failure(Exception(rpcResponse.error.message))
            } else if (rpcResponse.result != null) {
                Result.success(rpcResponse.result)
            } else {
                Result.failure(Exception("No result in response"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Get lockdown status
     */
    fun getLockdownStatus(): Result<LockdownStatus> {
        return rpc("lockdown.status")
    }

    /**
     * Get a bonding challenge
     */
    fun getChallenge(): Result<Challenge> {
        return rpc("lockdown.get_challenge")
    }

    /**
     * Claim ownership (admin bonding)
     */
    fun claimOwnership(
        displayName: String,
        deviceName: String,
        deviceFingerprint: String,
        passphraseCommitment: String,
        challengeResponse: String? = null
    ): Result<BondingResponse> {
        val params = mutableMapOf(
            "display_name" to displayName,
            "device_name" to deviceName,
            "device_fingerprint" to deviceFingerprint,
            "passphrase_commitment" to passphraseCommitment
        )
        challengeResponse?.let { params["challenge_response"] = it }
        return rpc("lockdown.claim_ownership", params)
    }

    /**
     * Transition lockdown mode
     */
    fun transitionMode(target: String): Result<Map<String, String>> {
        return rpc("lockdown.transition", mapOf("target" to target))
    }

    /**
     * Get security categories
     */
    fun getSecurityCategories(): Result<List<DataCategory>> {
        return rpc("security.get_categories")
    }

    /**
     * Set security category permission
     */
    fun setSecurityCategory(category: String, permission: String): Result<Map<String, Boolean>> {
        return rpc("security.set_category", mapOf(
            "category" to category,
            "permission" to permission
        ))
    }

    /**
     * Get security tiers
     */
    fun getSecurityTiers(): Result<List<SecurityTier>> {
        return rpc("security.get_tiers")
    }

    /**
     * Set security tier
     */
    fun setSecurityTier(tier: String): Result<Map<String, Boolean>> {
        return rpc("security.set_tier", mapOf("tier" to tier))
    }

    /**
     * List devices
     */
    fun listDevices(): Result<List<Device>> {
        return rpc("device.list")
    }

    /**
     * Approve device
     */
    fun approveDevice(deviceId: String, approvedBy: String): Result<Map<String, Boolean>> {
        return rpc("device.approve", mapOf(
            "device_id" to deviceId,
            "approved_by" to approvedBy
        ))
    }

    /**
     * Reject device
     */
    fun rejectDevice(deviceId: String, rejectedBy: String, reason: String): Result<Map<String, Boolean>> {
        return rpc("device.reject", mapOf(
            "device_id" to deviceId,
            "rejected_by" to rejectedBy,
            "reason" to reason
        ))
    }

    /**
     * Create invite
     */
    fun createInvite(role: String, expiration: String, maxUses: Int, createdBy: String): Result<Invite> {
        return rpc("invite.create", mapOf(
            "role" to role,
            "expiration" to expiration,
            "max_uses" to maxUses.toString(),
            "created_by" to createdBy
        ))
    }

    /**
     * List invites
     */
    fun listInvites(): Result<List<Invite>> {
        return rpc("invite.list")
    }

    /**
     * Revoke invite
     */
    fun revokeInvite(inviteId: String, revokedBy: String): Result<Map<String, Boolean>> {
        return rpc("invite.revoke", mapOf(
            "invite_id" to inviteId,
            "revoked_by" to revokedBy
        ))
    }

    /**
     * Generate setup QR
     */
    fun generateSetupQR(): Result<Map<String, String>> {
        return rpc("qr.generate_setup")
    }

    //region Push Notification Methods

    @Serializable
    data class PushTokenResponse(
        val success: Boolean,
        val message: String,
        val device_id: String? = null
    )

    /**
     * Register FCM push token with the bridge
     */
    fun registerPushToken(deviceId: String, token: String, platform: String): Result<PushTokenResponse> {
        return rpc("push.register_token", mapOf(
            "device_id" to deviceId,
            "token" to token,
            "platform" to platform
        ))
    }

    /**
     * Unregister FCM push token from the bridge
     */
    fun unregisterPushToken(deviceId: String, token: String): Result<PushTokenResponse> {
        return rpc("push.unregister_token", mapOf(
            "device_id" to deviceId,
            "token" to token
        ))
    }

    //endregion
}
